pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -910,18 +914,40 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
 		reStrings := make([]string, 0, len(l7Rules.DNS))
 		for _, dnsRule := range l7Rules.DNS {
 			if len(dnsRule.MatchName) > 0 {
-				dnsRuleName := strings.ToLower(dns.Fqdn(dnsRule.MatchName))
-				dnsPatternAsRE := matchpattern.ToRegexp(dnsRuleName)
-				reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+				dnsRuleName := FromFqdn(dnsRule.MatchName)
+				dnsPatternAsRE, _ := CustomToRegexp(Reverse(dnsRuleName))
+				reStrings = append(reStrings, dnsPatternAsRE)
```

Comment:
Do the `matchName`s need to be reversed? From what I understand, reversing the string makes sense if it's a pattern which contains multi-char matches such as `*` or `.+` type matches, i.e. `*.*.svc.cluster.local` because that allows the regex engine to "exit early". To elaborate, every DNS name comparison against the regex that starts with `*` and `*` will force the engine to go through all possible matches upfront, rather than when necessary.

For example, take the pattern `*.*.svc.cluster.local` and a DNS name `foo.bar.svc.cluster.local`. Without reversing, the regex engine will first process `foo` and `bar` and they'll match against the corresponding `*`'s. However, if the DNS name instead was `foo.bar.something.else.local`, then only after the `*` checks, does the regex engine realize this DNS name doesn't match.

Let's say we take the reverse of the pattern `local.cluster.svc.*.*` and `foo.bar.something.else.local`. The first `local` part will pass as a basic string match (less expensive than wildcard presumably). The second part, `else`, matches against `cluster` and the engine returns false, whereas if it was `*` it would've return true at this part of the domain.

If it's a name, then I don't see any improvement related to the above which can be possible, so we are probably just wasting cycles reversing the string.

(I think it would be helpful to put this as a code comment if accurate to explain why we do this.)